### ISSUE TITLE
[Uploader] Make block data upload robust against failures

### ISSUE THREAD

**Issue by smnzhu** [onflow/flow-go#2273]
Currently if EN is restarted or crashes after a block gets executed but before the block data is successfully uploaded to GCP, then upon restart that block won’t get uploaded.

Currently there is no retry mechanism, because it would be rather hard to do - it requires trie updates which aren't easily obtained outside of execution window.

Perhaps we can persist the block data to disk before marking the execution of the block as complete. This way we have it saved in case we need to retry, but at the same time we don’t need to wait for the upload to complete, this part can still happen async.

**Comment by j1010001** [comment]
Update after chat with Maks:
When EN restarts it tries to catch up and execute blocks. It loads them into a memory and then start executing
but if there are a lots of blocks, they won't fit into the memory. Ideally, we would have some form of a buffer or list - load blocks from one end, and execute from the other, so it makes progress. 
In a recent issue on testnet where we had to re0execute blocks we patched EN to load 100k blocks, execute and crash. We kept restarting the EN until it executed and uploaded all blocks.
